src/wallet/wallet.cpp

```diff
@@ -3284,7 +3285,7 @@ bool CWallet::Lock()
         return false;
 
     {
-        LOCK(cs_wallet);
+        LOCK2(m_relock_mutex, cs_wallet);
```

Comment:
Was surprised to see that this wasn't causing any lock ordering deadlock. Because, for example, `ChangeWalletPassphrase` first locks `cs_wallet` and then call `Lock()` (which locks `m_relock_mutex` first, which is the opposite lock ordering direction).

So.. got deeper over it and was able to trigger the deadlock detection on the GUI "change wallet passphrase" workflow.
You can easily replicate it by enabling the debug mode (--enable-debug at configure time) and try to change the wallet passphrase there.

This is not happening on the RPC command merely because this two recursive locks are acquired at the `ChangeWalletPassphrase` caller, which locks them in the proper order (which does not happen on the GUI's `changeWalletPassphrase` interface method).

Reply:
Good catch, I was able to replicate this locally. Should be fixed now.